celery/canvas.py

```diff
@@ -1170,21 +1170,29 @@ def _apply_tasks(self, tasks, producer=None, app=None, p=None,
             # we are able to tell when we are at the end by checking if
             # next_task is None.  This enables us to set the chord size
             # without burning through the entire generator.  See #3021.
+            group_ids = set()
+            chord_size = 0
             for task_index, (current_task, next_task) in enumerate(
                 lookahead(tasks)
             ):
+                # We expect that each task must be part of the same group which
+                # seems sensible enough. If that's somehow not the case we'll
+                # end up messing up chord counts and there are all sorts of
+                # awful race conditions to think about, we perform a sanity
+                # check for this behaviour at least.
                 sig, res, group_id = current_task
-                _chord = sig.options.get("chord") or chord
-                if _chord is not None and next_task is None:
-                    chord_size = task_index + 1
-                    if isinstance(sig, _chain):
-                        if sig.tasks[-1].subtask_type == 'chord':
-                            chord_size = sig.tasks[-1].__length_hint__()
-                        else:
-                            chord_size = task_index + len(sig.tasks[-1])
+                group_ids.add(group_id)
+                chord_obj = sig.options.get("chord") or chord
+                # We need to check the chord size of each contributing task so
+                # that when we get to the final one, we can correctly set the
+                # size in the backend and the chord can be sensible completed.
+                chord_size += _chord._descend(sig)
+                if chord_obj is not None and next_task is None:
```

Comment:
If this is a sanity check, please add `if __debug__` so that the entire check would not be executed in production.

Reply:
This bit isn't but the assertion below is. I'll make sure that only happens `if __debug__` but build the `group_ids` set as usual and just not use it when `not __debug__`

Reply:
This is done now. I'll resolve the other thread and leave this one for you.

Reply:
asserts are already removed when `not __debug__`.

There's no reason to execute the `if chord_obj is not None and next_task is None` condition and there's no reason to declare or to add group ids to the set other than debugging purposes as far as I can see.

Reply:
Oh, I see what you mean. We do need to execute that condition to check that we're actually at the final task in the iterable, because we don't set the actual chord size in the backend until we've exhausted the iterable created from `enumerate(lookahead(tasks))`, which is indicated by `next_task is None` and done when that conditional is truthy.

As a note, I have NFI if there's a race condition there with the chord being completed early because we don't set the final size until some later point and perhaps tasks which this code fires off start landing. The potential issue crossed my mind but I haven't had time to think about it much and this PR is mostly aimed to avoid the chord size mis-calculation.

I put the group IDs in a set mostly because it felt like a low cost sanity check. I'll probably just remove it rather than committing a bunch of `if __debug__` conditionals. Thoughts?

Reply:
It is a low-cost sanity check if there aren't many groups.
Since we can't really expect what our users will do with the code, we shouldn't assume how large a canvas is.

Reply:
I'll just trim it out then, and push an amended changeset shortly. :)

Reply:
This has been removed now